File: primatis-api/src/main/java/be/primatis/catalogue/dto/UpdateCopyRequest.java
package be.primatis.catalogue.dto;

import be.primatis.catalogue.CopyCondition;
import com.fasterxml.jackson.annotation.JsonIgnore;

/**
 * Contrat REST de modification staff d'un {@code Copy} (DEV-06.6,
 * {@code COPY_MANAGE}). Même mécanisme exact de PATCH sparse que
 * {@link UpdateGenreRequest}/{@link UpdateAuthorRequest}/{@link
 * UpdateTitleRequest} : classe JavaBean, indicateur de présence par champ.
 *
 * <p>Volontairement absent : {@code availabilityStatus} (action dédiée
 * {@link UpdateCopyAvailabilityRequest}, {@code PATCH .../availability}) et
 * {@code titleId} (immuable dans DEV-06.6 — un Copy ne change pas de Title,
 * aucun transfert d'exemplaire implémenté).
 *
 * <p>{@code inventoryCode} : présent+{@code null}/blanc refusé (colonne
 * {@code NOT NULL}/{@code UNIQUE}). {@code location} (nullable) :
 * présent+{@code null} efface. {@code copyCondition} : présent+{@code null}
 * refusé (colonne {@code NOT NULL}) ; si la nouvelle valeur est {@code LOST}
 * ou {@code OUT_OF_SERVICE}, le Service impose {@code UNAVAILABLE} dans la
 * même transaction (invariant physique déjà FIGÉ) — sinon
 * {@code availabilityStatus} n'est jamais modifié par ce endpoint, y compris
 * s'il vaut déjà {@code ON_LOAN}/{@code RESERVED} (DEV-06.6 §30 : un PATCH
 * {@code inventoryCode}/{@code location}/{@code copyCondition=GOOD|DAMAGED}
 * ne touche jamais un {@code ON_LOAN}/{@code RESERVED} existant).
 */
public class UpdateCopyRequest {

    private String inventoryCode;
    private boolean inventoryCodePresent;

    private String location;
    private boolean locationPresent;

    private CopyCondition copyCondition;
    private boolean copyConditionPresent;

    public UpdateCopyRequest() {
    }

    public String getInventoryCode() {
        return inventoryCode;
    }

    public void setInventoryCode(String inventoryCode) {
        this.inventoryCode = inventoryCode;
        this.inventoryCodePresent = true;
    }

    @JsonIgnore
    public boolean isInventoryCodePresent() {
        return inventoryCodePresent;
    }

    public String getLocation() {
        return location;
    }

    public void setLocation(String location) {
        this.location = location;
        this.locationPresent = true;
    }

    @JsonIgnore
    public boolean isLocationPresent() {
        return locationPresent;
    }

    public CopyCondition getCopyCondition() {
        return copyCondition;
    }

    public void setCopyCondition(CopyCondition copyCondition) {
        this.copyCondition = copyCondition;
        this.copyConditionPresent = true;
    }

    @JsonIgnore
    public boolean isCopyConditionPresent() {
        return copyConditionPresent;
    }
}

File: primatis-web/src/app/catalogue/models/update-copy-request.ts
import { CopyCondition } from './copy-condition';

/**
 * Voir `be.primatis.catalogue.dto.UpdateCopyRequest` côté backend. Même
 * mécanisme de PATCH sparse à trois états que `UpdateTitleRequest`.
 *
 * Volontairement absents : `availabilityStatus` (action dédiée
 * `UpdateCopyAvailabilityRequest`, `PATCH .../availability`) et `titleId`
 * (immuable — un Copy ne change pas de Title).
 *
 * - `inventoryCode` : `null`/vide explicite est rejeté par le backend
 *   (colonne `NOT NULL`/`UNIQUE`) — pas de `| null` ici.
 * - `location` : `null` est accepté et efface la valeur — `| null`
 *   légitime.
 * - `copyCondition` : `null` explicite est rejeté par le backend (colonne
 *   `NOT NULL`) — PAS de `| null` ici (à ne pas confondre avec le type
 *   suggéré `CopyCondition | null` : le backend refuse `copyCondition`
 *   présent+`null` avec `COPY_CONDITION_MUST_NOT_BE_NULL`, vérifié dans
 *   `UpdateCopyRequest.java`). Si la nouvelle valeur est `LOST` ou
 *   `OUT_OF_SERVICE`, le backend impose `UNAVAILABLE` dans la même
 *   transaction — non dupliqué ici.
 */
export interface UpdateCopyRequest {
  inventoryCode?: string;
  location?: string | null;
  copyCondition?: CopyCondition;
}
